spring-kafka/src/test/java/org/springframework/kafka/listener/DeliveryAttemptAwareRetryListenerIntegrationTests.java

```diff
@@ -0,0 +1,280 @@
+
+	private boolean awaitLatch(CountDownLatch latch) {
+		try {
+			return latch.await(60, TimeUnit.SECONDS);
+		}
+		catch (Exception e) {
+			fail(e.getMessage());
+			throw new RuntimeException(e);
+		}
+	}
+
+	private static CommonErrorHandler createErrorHandler(int interval, int maxAttemptCount) {
+		final FixedBackOff fixedBackOff = new FixedBackOff(interval, maxAttemptCount);
+		DefaultErrorHandler errorHandler = new DefaultErrorHandler(fixedBackOff);
+		errorHandler.setRetryListeners(new DeliveryAttemptAwareRetryListener());
+		return errorHandler;
+	}
+
+	private static ConcurrentKafkaListenerContainerFactory<String, String> createListenerContainerFactory(
+			ConsumerFactory<String, String> consumerFactory, CommonErrorHandler errorHandler) {
+
+		ConcurrentKafkaListenerContainerFactory<String, String> factory = new ConcurrentKafkaListenerContainerFactory<>();
+		factory.setConsumerFactory(consumerFactory);
+		factory.setCommonErrorHandler(errorHandler);
+
+		final ContainerProperties containerProperties = factory.getContainerProperties();
+		containerProperties.setDeliveryAttemptHeader(true);
+		return factory;
+	}
+
+	static class TestTopicListener0 {
+		final List<Header> receivedHeaders = new ArrayList<>();
+
+		@KafkaListener(
+				topics = TEST_TOPIC0,
+				containerFactory = MAIN_TOPIC_CONTAINER_FACTORY0,
+				batch = "true")
+		public void listen(List<ConsumerRecord<?, ?>> records) {
+			for (ConsumerRecord<?, ?> record : records) {
+				Iterable<Header> headers = record.headers().headers(KafkaHeaders.DELIVERY_ATTEMPT);
+				for (Header header : headers) {
+					receivedHeaders.add(header);
+				}
+			}
+			latch0.countDown();
+			throw new RuntimeException("Failed.");
+		}
+	}
+
+	static class TestTopicListener1 {
+		final List<Header> receivedHeaders = new ArrayList<>();
+
+		@KafkaListener(
+				topics = TEST_TOPIC1,
+				containerFactory = MAIN_TOPIC_CONTAINER_FACTORY1,
+				batch = "true")
+		public void listen(List<ConsumerRecord<?, ?>> records) {
+			for (ConsumerRecord<?, ?> record : records) {
+				Iterable<Header> headers = record.headers().headers(KafkaHeaders.DELIVERY_ATTEMPT);
+				for (Header header : headers) {
+					receivedHeaders.add(header);
+				}
+			}
+			latch1.countDown();
+			throw new RuntimeException("Failed.");
+		}
+	}
+
+	@Configuration
+	static class TestConfiguration {
+
+		@Bean
+		TestTopicListener0 testTopicListener0() {
+			return new TestTopicListener0();
+		}
+
+		@Bean
+		TestTopicListener1 testTopicListener1() {
+			return new TestTopicListener1();
+		}
+	}
+
+	@Configuration
+	static class KafkaProducerConfig {
+
+		@Autowired
+		EmbeddedKafkaBroker broker;
+
+		@Bean
+		ProducerFactory<String, String> producerFactory() {
+			Map<String, Object> props = KafkaTestUtils.producerProps(
+					this.broker.getBrokersAsString());
+			props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+			props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
```

Comment:
The `value.serializer` is a `StringSerializer` by default in the `KafkaTestUtils.producerProps()`.
And why do we really need a `StringSerializer` for key in these tests?
Why cannot we just rely on the defaults from the `KafkaTestUtils.producerProps()`?